rfcs/12-primitive-shapes.md

```diff
@@ -0,0 +1,538 @@
+
+Meshing could be naturally extended with other libraries or parameters. For example, a sphere by default might use `Icosphere`, but could be extended to generate UV spheres or quad spheres. For 2d types, we can use a crate such as `lyon` to generate 2d meshes from parameterized primitive shapes within the `Meshable` interface.
+
+### Bounding
+
+For use in spatial queries, broad phase collision detection, and raycasting, `Bounding` and `Bounding2d` traits are provided. These traits are implemented for types that define position in addition to the shape and size defined in the base primitive. The basic functionality of this trait is to check whether one bounding shape is contained within another bounding shape.
+
+### Colliders
+
+Colliders can provide intersection checks against other colliders, as well as check if they are within a bounding volume.
+
+### 3D and 2D
+
+This RFC provides independent 2d and 3d primitives. Recall that the purpose of this is to provide lightweight types, so there are what appear to be duplicates in 2d and 3d, such as `Line` and `Line2d`. Note that the 2d version of a line is smaller than its 3d counterpart because it is only defined in 2d. 3d geometry (or 2d with depth) is assumed to be the default for most cases. The names of the types were chosen with this in mind.
+
+## Implementation strategy
+
+### Helper Types
+
+```rust
+/// Stores an angle in radians, and supplies builder functions to prevent errors (from_radians, from_degrees)
+struct Angle(f32);
+```
+### Traits
+
+These traits are provided as reference to illustrate how these might be used. The details of implementation and interface should be determined in a separate RFC, PR, or independent prototypes.
+
+```rust
+trait Meshable{
+  fn mesh(&self) -> Mesh;
+};
+
+trait Bounding {
+  fn within(&self, other: &impl Bounding) -> bool;
+  fn contains(&self, collider: &impl Collider) -> bool;
+}
+
+trait Bounding2d {
+  fn within(&self, other: &impl Bounding2d) -> bool;
+  fn contains(&self, collider: &impl Collider2d) -> bool;
+}
+
+trait Collider {
+  fn collide(&self, other: &impl Collider) -> Option(Collision);
+  fn within(&self, bounds: &impl Bounding) -> bool;
+}
+
+trait Collider2d {
+  fn collide(&self, other: &impl Collider2d) -> Option(Collision2d);
+  fn within(&self, bounds: &impl Bounding2d) -> bool;
+}
+```
+
+### 3D Geometry Types
+
+```rust
+struct Point(Vec3)
+
+/// Vector direction in 3D space that is guaranteed to be normalized through its getter/setter.
+struct Direction(Vec3)
+impl Meshable for Direction {}
+
+struct Plane {
+  point: Point,
+  normal: Direction,
+}
+impl Meshable for Plane {}
```

Comment:
I'm not sure an infinite Plane should implement Meshable. Maybe we should have a separate bounded Plane type?

Similarly, Ray, Line, LineSegment etc don't make sense as Meshable without more details such as width, etc.

Reply:
What about the infinite reverse frustum? :)

Reply:
A few things to consider here:

- The meshing trait is demonstrative, actual implementation is out of scope of this RFC
- The intent is to provide some trait that you can use to very easily visually debug your primitive geometry. This interface would allow you to simply iterate over anything that implements `Meshable`, and spit out a mesh. Maybe it would be better to call the trait `DebugMesh`, but again, it's purely demonstrative.